### DISCUSSION TITLE
unable to resolve crossref for python chunks in quarto

### DISCUSSION THREAD

**Discussion by KipBalkcom-USDA** [quarto-dev/quarto-cli#7667] [Q&A] [answered]
### Description

I was using a quarto document to render to a word document that contained r and python code chunks with cross references in the text that referred to table or figure numbers.  Everything had worked perfectly, but ...

I attempted to put inline output from python into the text before I realized this only worked for r code.  

In the process of trying different options, nothing negative happened immediately. However,  I eventually received a WARNING: unable to resolve crossref @tbl-name for all the python chunks I was using that contained cross references. All r chunks behave as they should. The output for the python chunks is present in the rendered document, but no cross ref works for the python chunks.

I am only showing a python chunk, but my document contains r code; therefore, I omitted the jupyter: python3 statement in the yaml.  There was no previous problem omitting that statement.

I don't know what happened or how to resolve. 

Please note you would have to change the attached file from .txt to .py

```` qmd
 ---
format:
  docx: 
keep-tex: false
execute: 
  echo: False
  warning: false
page-width: 7.5
tbl-colwidths: auto
always_allow_html: true
editor_options: 
  chunk_output_type: inline
---

```{python}
# | label: python libraries

from pyprojroot.here import here

```  

```{python, file=here("n_treatments.py")}
# | label: tbl-rates
# | tbl-cap: Just a caption.

```
Just a statement to reference the values shown in @tbl-rates.

[n_treatments.txt](https://github.com/quarto-dev/quarto-cli/files/13431270/n_treatments.txt)

````

**Comment by mcanouil** [comment]
From what you are saying you are using knitr and reticulate, is it correct?
Which version of Quarto are you using?

You seem to not have closed properly your example as the attachment is included in it.

~~I have difficulties to understand what a Python script has to do with Quarto as you said to use 'py' extension.~~
Edit: oh you are using knitr to source an external Python script.

Note that if your document can be shared as a flat file which seems to be the case, you could simply paste that here.
If you could also remove all non Quarto options and options that has nothing to do with your issue that would help.
Currently the code you shared as plain text is invalid as it mixes RMarkdown and Quarto options.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by KipBalkcom-USDA** [reply]
I was trying to provide the python file that was being referenced in the quarto document.  I did not intend to attach it as it did.  I am not sure what you mean by non Quarto options.
[n_treatments.txt](https://github.com/quarto-dev/quarto-cli/files/13431926/n_treatments.txt)

**Reply by mcanouil** [reply, marked as answer]
You have many typos and unnecessary options at least in your example.
- YAML style option for R/Python code cell is `#|` not `# |`.
- You should not define key without value, *i.e.*, `docx: `.
- `keep-tex` is not relevant for `docx`.

if you want to use knitr/reticulate for Python code without R code cells, as said by Carlos, you need to explicitly set the engine.

Fixing everything and make it small works just fine:

````qmd
---
format: docx
engine: knitr
---

```{python}
#| label: python libraries
from pyprojroot.here import here
```  

```{python}
#| label: tbl-rates
#| tbl-cap: Just a caption.
#| file: n_treatments.py
```

Just a statement to reference the values shown in @tbl-rates.
````

```py
import pandas as pd
from tabulate import tabulate
from IPython.display import Markdown

# define treatments
n_trts = pd.DataFrame({
    "Treatment": [1, 2, 3, 4, 5, 6],
    "Fall N (kg ha^-1^)": [0, 34] * 3,
})

Markdown(tabulate(
    n_trts, headers="keys", showindex=False, numalign="center", tablefmt="pipe"))
```

**Reply by KipBalkcom-USDA** [reply]
Thank you very much.  I am embarrassed, but it was the typos.  When I corrected them, everything works as it did before.  I obviously introduced those typos when I was trying to make inline python code work in quarto.

**Comment by cscheid** [comment]
Code cell syntax with non-yaml is knitr-specific syntax. That `file=` syntax is only supported in the knitr engine.

You can verify that there's no output happening by:

- inspecting the .md file in `keep-md: true`, or 
- making `n_treatments.py` crash on purpose and verify it doesn't happen at all when jupyter is the engine, or
- adding `engine: knitr` and verify the output appears as expected.

With no `r` code cells, the auto-detection logic switches to the jupyter engine, and so you get the behavior you're seeing.

**Reply by KipBalkcom-USDA** [reply]
The word document is rendered correctly with the output from the python chunks, but the cross-ref does not work on the python chunks and provides the warning I mentioned.  I see I should have put an r chunk in the example now so my example does not switch to the jupyter engine.

**Reply by KipBalkcom-USDA** [reply]
Sorry, I forgot to provide this.
Quarto: 1.3.450
windows 10
python: 3.12
R: 4.3.0
